File: ios/GradeThread/Inventory/InventoryDerivation.swift
import Foundation

/// US-1017: main-thread memoization for the inventory screen's three expensive
/// derived collections — the filtered+sorted list, the per-stage counts, and
/// the filter-sheet facet index.
///
/// SwiftUI re-evaluates a view's `body` on *every* `@State` change, including
/// ones that have nothing to do with filtering (presenting the sync sheet,
/// toggling select mode, a transient refresh banner). Any computed property
/// read during that pass is recomputed from scratch. On a large catalog the
/// full `InventoryFilter.apply` (O(n) match + O(n log n) sort), the per-stage
/// count pass (O(n · stages)), and `InventoryFacets.derive` (O(n) + several
/// sorts) are all wasted work when their inputs haven't changed.
///
/// This cache stores the last result of each derivation keyed on a cheap
/// integer signature of its inputs; a getter recomputes only when that
/// signature changes. It is held in `@State` as a reference type so the cache
/// survives the value-type view being re-created across renders, and — because
/// it is a plain `final class`, NOT `@Observable`/`ObservableObject` — mutating
/// the cache during `body` does not itself invalidate the view (no render
/// loop).
///
/// The `*PassCount` properties are instrumentation: tests assert a derivation
/// runs at most once for a stable key (and once more per genuine input change),
/// which is exactly the "doesn't recompute on unrelated re-renders" contract.
final class InventoryDerivation {

    // MARK: - Memo slots

    private var filteredKey: Int?
    private var filteredValue: [LocalInventoryItem] = []

    private var stageCountsKey: Int?
    private var stageCountsValue: [InventoryStage: Int] = [:]

    private var facetsKey: Int?
    private var facetsValue: InventoryFacets = .empty

    // US-1520: item ids with at least one cached photo, for the photo facet.
    private var photoItemIdsKey: Int?
    private var photoItemIdsValue: Set<String> = []

    // MARK: - Instrumentation (for tests)

    /// Number of times the full filter+sort pass actually executed.
    private(set) var filterPassCount = 0
    /// Number of times the per-stage count pass actually executed.
    private(set) var stageCountsPassCount = 0
    /// Number of times the facet derive actually executed.
    private(set) var facetsPassCount = 0

    // MARK: - Memoized getters

    /// Returns the cached filtered list, recomputing via `compute` only when
    /// `key` differs from the last call.
    func filteredItems(key: Int, _ compute: () -> [LocalInventoryItem]) -> [LocalInventoryItem] {
        if filteredKey != key {
            filteredKey = key
            filteredValue = compute()
            filterPassCount += 1
        }
        return filteredValue
    }

    /// Returns the cached per-stage counts, recomputing only on `key` change.
    func stageCounts(key: Int, _ compute: () -> [InventoryStage: Int]) -> [InventoryStage: Int] {
        if stageCountsKey != key {
            stageCountsKey = key
            stageCountsValue = compute()
            stageCountsPassCount += 1
        }
        return stageCountsValue
    }

    /// Returns the cached facet index, recomputing only on `key` change. The
    /// key folds in the item signature *and* the source-name map, so renaming a
    /// source (which changes only labels) still refreshes the facet.
    func facets(key: Int, _ compute: () -> InventoryFacets) -> InventoryFacets {
        if facetsKey != key {
            facetsKey = key
            facetsValue = compute()
            facetsPassCount += 1
        }
        return facetsValue
    }

    /// US-1520: cached photo-bearing item id set (one `LocalItemPhoto` id pass),
    /// recomputing only on `key` change. Feeds the photo facet so
    /// ``InventoryFilter/matches`` never faults each item's lazy `photos`
    /// relationship. Same staleness contract as the filter memo: the items
    /// signature gates it, and a photo change reaches the list via the same
    /// signals that already refresh the filtered list.
    func photoItemIds(key: Int, _ compute: () -> Set<String>) -> Set<String> {
        if photoItemIdsKey != key {
            photoItemIdsKey = key
            photoItemIdsValue = compute()
        }
        return photoItemIdsValue
    }

    // MARK: - Cache keys (pure, shared with the view + unit-testable)

    /// Cheap content signature of an item snapshot — `count` folded with each
    /// row's `id` + `updatedAt`. Two snapshots with identical (id, updatedAt)
    /// pairs hash equal, so an unrelated `body` re-render (same cache) reuses the
    /// memo; any edit bumps `updatedAt` and invalidates it. O(n) but far cheaper
    /// than the passes it gates. The per-process `Hasher` seed is stable within a
    /// run, so signatures are only ever compared against others built in the
    /// same process.
    static func itemsSignature(_ items: [LocalInventoryItem]) -> Int {
        var hasher = Hasher()
        hasher.combine(items.count)
        for item in items {
            hasher.combine(item.id)
            hasher.combine(item.updatedAt)
        }
        return hasher.finalize()
    }

    /// Key for the filtered+sorted list. Folds in exactly the inputs
    /// ``InventoryFilter/apply(_:stage:search:sort:criteria:now:soldDates:serverSearchIds:)``
    /// reads — nothing else — so an unrelated `@State` change leaves it
    /// unchanged and the memo holds.
    static func filterKey(
        itemsSignature: Int,
        stage: InventoryStage,
        query: String,
        sort: SortOption,
        criteria: InventoryFilterCriteria,
        soldDates: [String: Date],
        serverSearchIds: Set<String>?,
        unlistedFilter: InventoryStage.UnlistedFilter = .all
    ) -> Int {
        var hasher = Hasher()
        hasher.combine(itemsSignature)
        hasher.combine(stage)
        hasher.combine(query)
        hasher.combine(sort)
        hasher.combine(criteria)
        hasher.combine(soldDates)
        hasher.combine(serverSearchIds)
        hasher.combine(unlistedFilter)
        return hasher.finalize()
    }

    /// Key for the facet index — the item signature plus the source-name map (a
    /// source rename changes labels only, but must still refresh the facet).
    static func facetsKey(itemsSignature: Int, sourceNames: [String: String]) -> Int {
        var hasher = Hasher()
        hasher.combine(itemsSignature)
        hasher.combine(sourceNames)
        return hasher.finalize()
    }
}

File: android/app/src/main/java/com/gradethread/app/inventory/InventoryDerivation.kt
package com.gradethread.app.inventory

import com.gradethread.app.sync.db.InventoryItemEntity

/**
 * US-1342: a keyed memo over the derivation pipeline (iOS
 * `InventoryDerivation`).
 *
 * The point is that the inventory screen has ~20 pieces of unrelated UI state
 * — a filter sheet opening, select mode toggling, a refresh banner — and each
 * of them recomposes the screen. Without this cache every one of those would
 * re-run a full O(n log n) filter + sort over the whole inventory.
 *
 * Each slot's key is built from EXACTLY the inputs its computation reads and
 * nothing else, so unrelated state cannot invalidate it. Hold one instance
 * per screen (`remember`), not per composition.
 *
 * Not thread-safe: it is read during composition on the main thread.
 */
class InventoryDerivation {

    private var filteredKey: Int? = null
    private var filteredValue: List<InventoryItemEntity> = emptyList()

    private var stageCountsKey: Int? = null
    private var stageCountsValue: Map<InventoryStage, Int> = emptyMap()

    private var facetsKey: Int? = null
    private var facetsValue: InventoryFacets = InventoryFacets()

    /**
     * Pass counters exist so tests can assert the memo contract directly —
     * "the value was right" is not the same claim as "it didn't recompute".
     */
    var filterPassCount: Int = 0
        private set
    var stageCountsPassCount: Int = 0
        private set
    var facetsPassCount: Int = 0
        private set

    // Every parameter is an input `apply` reads; the count is the filter's arity.
    @Suppress("LongParameterList")
    fun filtered(
        items: List<InventoryItemEntity>,
        stage: InventoryStage,
        query: String,
        sort: SortOption,
        criteria: InventoryFilterCriteria,
        photoItemIds: Set<String>? = null,
        serverSearchIds: Set<String>? = null,
        nowMillis: Long = System.currentTimeMillis(),
        unlistedFilter: UnlistedFilter = UnlistedFilter.ALL,
    ): List<InventoryItemEntity> {
        val key = filterKey(
            itemsSignature(items),
            stage,
            query,
            sort,
            criteria,
            photoItemIds,
            serverSearchIds,
            unlistedFilter,
        )
        if (key != filteredKey) {
            filteredKey = key
            filterPassCount++
            filteredValue = InventoryFilter.apply(
                items = items,
                stage = stage,
                query = query,
                sort = sort,
                criteria = criteria,
                photoItemIds = photoItemIds,
                serverSearchIds = serverSearchIds,
                nowMillis = nowMillis,
                unlistedFilter = unlistedFilter,
            )
        }
        return filteredValue
    }

    /** Counts every user-facing stage in one pass; ALL counts everything. */
    fun stageCounts(items: List<InventoryItemEntity>): Map<InventoryStage, Int> {
        val key = itemsSignature(items)
        if (key != stageCountsKey) {
            stageCountsKey = key
            stageCountsPassCount++
            val counts = mutableMapOf<InventoryStage, Int>()
            for (stage in InventoryStage.userFacing) counts[stage] = 0
            for (item in items) {
                for (stage in InventoryStage.userFacing) {
                    if (stage.matches(item.status)) counts[stage] = (counts[stage] ?: 0) + 1
                }
            }
            stageCountsValue = counts
        }
        return stageCountsValue
    }

    fun facets(items: List<InventoryItemEntity>, sourceNames: Map<String, String> = emptyMap()): InventoryFacets {
        // sourceNames is in the key because a source RENAME changes labels
        // without changing any item, and the facet list must still refresh.
        val key = 31 * itemsSignature(items) + sourceNames.hashCode()
        if (key != facetsKey) {
            facetsKey = key
            facetsPassCount++
            facetsValue = InventoryFacetsBuilder.derive(items, sourceNames)
        }
        return facetsValue
    }

    companion object {
        /**
         * Folds count + each row's id and updatedAt, so an EDIT invalidates
         * (updatedAt moves) but a re-render with the same rows does not.
         *
         * Never persist this — it is only ever compared against another
         * signature computed in the same process.
         */
        fun itemsSignature(items: List<InventoryItemEntity>): Int {
            var hash = items.size
            for (item in items) {
                hash = 31 * hash + item.id.hashCode()
                hash = 31 * hash + item.updatedAt.hashCode()
            }
            return hash
        }

        /** Exactly the eight inputs `apply` reads — nothing else. */
        @Suppress("LongParameterList")
        fun filterKey(
            itemsSignature: Int,
            stage: InventoryStage,
            query: String,
            sort: SortOption,
            criteria: InventoryFilterCriteria,
            photoItemIds: Set<String>?,
            serverSearchIds: Set<String>?,
            unlistedFilter: UnlistedFilter = UnlistedFilter.ALL,
        ): Int {
            var hash = itemsSignature
            hash = 31 * hash + stage.hashCode()
            hash = 31 * hash + query.hashCode()
            hash = 31 * hash + sort.hashCode()
            hash = 31 * hash + criteria.hashCode()
            hash = 31 * hash + (photoItemIds?.hashCode() ?: 0)
            hash = 31 * hash + (serverSearchIds?.hashCode() ?: 0)
            hash = 31 * hash + unlistedFilter.hashCode()
            return hash
        }
    }
}
